---
title: "RNA-seq pipeline (single-end reads)"
author: "LLi"
date: "7/12/2019"
output: html_document
---
<style type="text/css">

body{ /* Normal  */
      font-size: 16px;
  }
td {  /* Table  */
  font-size: 14px;
}
h1.title {
  font-size: 28px;
}
h1 { /* Header 1 */
  font-size: 24px;
  color: DarkBlue;
}
h2 { /* Header 2 */
    font-size: 20px;
}
code.r{ /* Code block */
    font-size: 12px;
}
pre { /* Code block - determines code spacing between lines */
    font-size: 14px;
}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE)
library(tidyverse)
library(knitr)
```

## Download fastq files from NCBI with SRR accession number
The RNA-seq data is from [*Ackermann AM et al.,*](https://www.ncbi.nlm.nih.gov/pubmed/26977395) "Integration of ATAC-seq and RNA-seq identifies human alpha cell and beta cell signature genes.", Mol Metab, 2016 Mar;5(3):233-244.

First download summary information regarding seq data by searching GEO accession number **GSE76268** in NCBI SRA website.

R:
```{r}
# Download Summary and RunInfo csv files of seq data from NCBI SRA website

# Read csv files in R
sra_result <- read.csv("~/seq_tutorials/sra_result.csv")
kable(sra_result[1:3, 1:6], caption = "sra_result.csv snapshot")

SraRunInfo <- read.csv("~/seq_tutorials/SraRunInfo.csv")
kable(SraRunInfo[1:3, 1:6], caption = "SraRunInfo.csv snapshot")

# Make two column names of these two csv files the same
names(sra_result)[names(sra_result) == "Sample.Accession"] <- "Sample"
names(sra_result)[names(sra_result) == "Experiment.Accession"] <- "Experiment"

# Merge the csv files by Sample and Experiment
GSE76268_merge <- merge(sra_result, SraRunInfo, by = c("Sample", "Experiment"))

# Remove columns with NA
GSE76268_runInfo <- GSE76268_merge[colSums(!is.na(GSE76268_merge)) > 0]

# Subset with columns that provides info to concatenate fastq files
cat_runInfo <- select(GSE76268_runInfo, c("Sample", "Run", "Experiment.Title"))
kable(cat_runInfo[1:29, ], caption = "Information for concatenating RNA-seq fastq files")
```
```{r eval=FALSE}
# Save csv files
write.csv(GSE76268_runInfo,'GSE76268_runInfo.csv')
write.csv(cat_runInfo, 'cat_runInfo.csv')
```

Then download SRA files to Linux server.

Shell:
```{bash eval=FALSE}
# Create a new directory for fastq files
mkdir input
cd input

# Download SRA files from NCBI (SRR3048055 to SRR3048083)
# Attention: SRA file download takes a lot of time, therefore run it in the background by screen or nohup
for (( i = 55; i <= 83; i++ ))
do
  echo "Start downloading SRR30480$i"
  wget https://sra-download.ncbi.nlm.nih.gov/traces/sra37/SRR/002976/SRR30480$i
done

# fastq-dump single-end seq data
for i in SRR*; do fastq-dump --gzip -A $i; done

# Merge files that belongs to the same sample
cat 

```





