---
title: "RSeQC demo"
author: "JWK & Islet group coding club meet-up"

output:
  ioslides_presentation: default
  powerpoint_presentation: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
library(knitr)
```

## Quality control checkpoints

QC should be applied to every step of RNA-seq: **raw sequencing reads**, **alignment** and **quantification**. 
The tools involved are:

- **FastQC** <https://www.bioinformatics.babraham.ac.uk/projects/fastqc/>
- **MultiQC** <https://multiqc.info/>
- **RSeQC** <http://rseqc.sourceforge.net/>
- **Picard**<https://broadinstitute.github.io/picard/>

## Output of aligner: SAM/BAM
The SAM (Sequence Alignment Map) file contains information for each individual read and its alignment to the genome. BAM is a binary, compressed version of the SAM file.

To view SAM/BAM file, type:
```{bash eval=FALSE}
file="your SAM/BAM file path"
samtools view -h $file | head -n 27
```

## SAM file header
|Tag|Description|
|--:|:----------|
|@HD| The header line| 
|@SQ| Reference sequence dictionary. The order of @SQ lines defines the alignment sorting order|
|VN| Format version|
|SN| Reference sequence name|
|SO| Sorting order of alignments|
|LN| Reference sequence length|

## SAM file alignment section
|No.	|Name	|Description|
|----:|:----|-----------|
|1	|QNAME|	Query NAME of the read or the read pair|
|2	|FLAG|	Bitwise FLAG (pairing, strand, mate strand, etc.)|
|3	|RNAME|	Reference sequence NAME|
|4	|POS|	1-Based leftmost POSition of clipped alignment|
|5	|MAPQ|	MAPping Quality (Phred-scaled)|

## SAM file alignment section

|No.	|Name	|Description|
|----:|:----|-----------|
|6	|CIGAR|	Extended CIGAR string (operations: MIDNSHP)|
|7	|MRNM|	Mate Reference NaMe (‘=’ if same as RNAME)|
|8	|MPOS|	1-Based leftmost Mate POSition|
|9	|ISIZE|	Inferred Insert SIZE|
|10	|SEQ|	Query SEQuence on the same strand as the reference|
|11	|QUAL|Query QUALity (ASCII-33=Phred base quality)|

## Insert size distribution (Picard)
The insert is normally the stretch of sequence between the paired-end adapters, but does not include adapter sequence.

```{r Pair-end reads, echo=FALSE, eval=TRUE, out.width = "280px"}
knitr::include_graphics("C:/Users/lingzili/Documents/seq_tutorials/plot/biostar_pair_end.png")
```

Only for pair-end library! There is no way to calculate average fragment size based on single-end sequencing data.

## Insert size distribution (Picard)

```{bash eval=FALSE}
picardPath="you file path to picard.jar"
file="your SAM/BAM file path"

# I: Input SAM or BAM file; O: File to write the output to; 
# H: File to write insert size Histogram to; 
# I: Discard any data that have fewer than this % of overall reads.

java -jar $picardPath CollectInsertSizeMetrics \
      I=$file \
      O=insert_size_metrics.txt \
      H=insert_size_histogram.pdf \
      INCLUDE_DUPLICATES=true \
      HISTOGRAM_WIDTH=800 \
      M=0.1
```

## Output of CollectInsertSizeMetrics

**insert_size_metrics.txt**

A data table that lists the read counts that corresponds to each insert size.

**insert_size_histogram.pdf**

```{r insert_histogram, echo=FALSE, eval=TRUE, out.width = "350px"}
knitr::include_graphics("C:/Users/lingzili/Documents/seq_tutorials/plot/SM4996_histo.png")
knitr::include_graphics("C:/Users/lingzili/Documents/seq_tutorials/plot/SM4998_histo.png")
```

## rRNA percentage
Ribosomal RNA (rRNA) often exceeding 90% of the total RNA. It is expected that prior library treatment of rRNA removal should result in less than 5% of the reads mapped to rRNA sequences. Excess rRNA (> 2%) will normally have to be filtered out so that differences in rRNA mapped reads across samples do not affect alignment rates and skew subsequent normalization of the data.


## Read distribution
It is expected that distribition profiles should be similar across samples. Less than 20% intronic mapped reads are expected in mRNA-seq libraries, whereas more than 40% in total RNA-seq libraries. A high intronic mapping suggests possible genomic DNA contamination and/or pre-mRNA. Around 55% of reads are normally observed mapping to exons.

## Gene body coverage
Uniform distributed reads across genebodies is a prerequisite for accurate quantification of gene expression. The line plots should look like this: "∩". mRNA-seq library preparation typically uses oligo-dT beads to capture mature transcripts and can be prone to 3' bias in genebody coverage if degraded RNA (RIN < 7) is used as input, which may introduce quantification bias.

