---
title: "RNA-seq mapping QC demo"
author: "JWK & Islet group coding club meet-up"

output:
  ioslides_presentation: default
  powerpoint_presentation: default
---
<style type="text/css">
code.r{
  font-size: 14px;
}
pre {
  font-size: 14px
}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
library(knitr)
library(tidyverse)
library(readxl)
library(scales)
```

## Quality control checkpoints

QC should be applied to every step of RNA-seq: **raw sequencing reads**, **alignment** and **quantification**. 
The tools involved are:

- **FastQC** <https://www.bioinformatics.babraham.ac.uk/projects/fastqc/>
- **MultiQC** <https://multiqc.info/>
- **RSeQC** <http://rseqc.sourceforge.net/>
- **Picard** <https://broadinstitute.github.io/picard/>

## Output of aligner: SAM/BAM file

- The SAM (Sequence Alignment Map) file contains information for **each individual read** and **its alignment to the genome**. 
- BAM is a binary, compressed version of the SAM file.

To view SAM/BAM file, type:
```{bash Samtools view, eval=FALSE}
file="your SAM/BAM file path"
samtools view -h $file | head -n 27
```

## SAM file header
|Tag|Description|
|--:|:----------|
|@HD| The header line| 
|@SQ| Reference sequence dictionary. The order of @SQ lines defines the alignment sorting order|
|VN| Format version|
|SN| Reference sequence name|
|SO| Sorting order of alignments|
|LN| Reference sequence length|

## SAM file alignment section
|No.	|Name	|Description|
|----:|:----|-----------|
|1	|QNAME|	Query NAME of the read or the read pair|
|2	|FLAG|	Bitwise FLAG (pairing, strand, mate strand, etc.)|
|3	|RNAME|	Reference sequence NAME|
|4	|POS|	1-Based leftmost POSition of clipped alignment|
|5	|MAPQ|	MAPping Quality (Phred-scaled)|

## SAM file alignment section

|No.	|Name	|Description|
|----:|:----|-----------|
|6	|CIGAR|	Extended CIGAR string (operations: MIDNSHP)|
|7	|MRNM|	Mate Reference NaMe (‘=’ if same as RNAME)|
|8	|MPOS|	1-Based leftmost Mate POSition|
|9	|ISIZE|	Inferred Insert SIZE|
|10	|SEQ|	Query SEQuence on the same strand as the reference|
|11	|QUAL|Query QUALity (ASCII-33=Phred base quality)|

## Insert size distribution (Picard)
The insert is normally the stretch of sequence between the paired-end adapters, but does not include adapter sequence.

```{r Pair-end reads, echo=FALSE, eval=TRUE, out.width = "280px"}
knitr::include_graphics("C:/Users/lingzili/Documents/seq_tutorials/plot/biostar_pair_end.png")
```

Only for pair-end library! There is no way to calculate average fragment size based on single-end sequencing data.

## Insert size distribution (Picard)

```{bash CollectInsertSizeMetrics, eval=FALSE}
# Define file path
locate picard.jar
picardPath="you file path to picard.jar"

file="your SAM/BAM file path"

# I: Input SAM or BAM file; O: File to write the output to; 
# H: File to write insert size Histogram to; 
# I: Discard any data that have fewer than this % of overall reads.

java -jar $picardPath CollectInsertSizeMetrics \
      I=$file \
      O=insert_size_metrics.txt \
      H=insert_size_histogram.pdf \
      INCLUDE_DUPLICATES=true \
      HISTOGRAM_WIDTH=800 \
      M=0.1
```

## Output of CollectInsertSizeMetrics

**insert_size_metrics.txt**

A data table that lists the **read counts** that corresponds to **each insert size**

**insert_size_histogram.pdf**

```{r insert_histogram, echo=FALSE, eval=TRUE, out.width = "350px"}
knitr::include_graphics("C:/Users/lingzili/Documents/seq_tutorials/plot/SM4996_histo.png")
knitr::include_graphics("C:/Users/lingzili/Documents/seq_tutorials/plot/SM4998_histo.png")
```

## rRNA percentage
Ribosomal RNA (rRNA) often exceeds 90% of the total RNA. Prior library treatment of rRNA removal should result in less than 5% of the reads mapped to rRNA sequences.

```{bash rRNA BAM, eval=FALSE}
# Define file path
locate split_bam
split_bam="your file path to split_bam.py"

file="your BAM file path" # BAM file should be sorted and indexed
rRNA_bed="your rRNA BED file path"
prefix_bam="prefix of your output BAM file"

python $split_bam -i $file -r $rRNA_bed -o $prefix_bam 1>>$prefix_bam.output 2>/dev/null

# Note: STDIN is represented by 0, STDOUT by 1, and STDERR by 2
```

## Ouput of split_bam.py

- 3 BAM files: in.bam, ex.bam, junk.bam
- Output file: contains info regarding reads in each category

|Total records:	|1110068	|
|--------------:|:--------|
|SM4998.in.bam (Reads consumed by input gene list):|     1056416|
|SM4998.ex.bam (Reads not consumed by input gene list):| 53652|
|SM4998.junk.bam (qcfailed, unmapped reads):|           0|

Excess rRNA (> 2%) will normally have to be filtered out! Because rRNA mapped reads can affect alignment rates and skew subsequent normalization of the data.

## Read genomic origin

- For mRNA-seq libraries, around 55% of reads are normally observed mapping to exons and less than 20% intronic mapped reads are expected
- For total RNA-seq libraries, more than 40% are observed mapping to introns
- Distribition profiles should be similar across samples
```{r Pie chart, echo=FALSE, fig.height=4, fig.width=4}
# Create a pie chart of reads genomic origin
SM4998_distribution <- read_excel("C:/Users/lingzili/Documents/seq_tutorials/data/SM4998.distribution.xlsx", col_types = c("text", "numeric", "numeric", "numeric", "numeric"))

subset_dist <- data.frame(SM4998_distribution[11:13, c(1, 5)])

blank_theme <- theme_minimal() +
  theme(
    axis.title.x = element_blank(),
    axis.title.y = element_blank(),
    axis.text.x = element_blank(),
    panel.border = element_blank(),
    panel.grid = element_blank(),
    axis.ticks = element_blank(),
    plot.title = element_text(size = 14, face = "bold")
  )

piePlot1 <- ggplot(subset_dist, aes(x = "", y = Per_Assigned_Tag, fill = Group))

piePlot1 + geom_bar(width = 1, stat = "identity") +
  coord_polar("y") +
  blank_theme +
  ggtitle("SM4998 (degraded)") +
  geom_text(aes(y = Per_Assigned_Tag / 6 + c(0, cumsum(Per_Assigned_Tag)[-length(Per_Assigned_Tag)]), 
                label = percent(Per_Assigned_Tag)), size = 4)
```

## read_distribution.py

```{bash Read distribution, eval=FALSE}
# Define file path
locate read_distribution
read_distribution="yout file path to read_distribution.py"

file="your SAM/BAM file path"
RefBED="your path to reference genome BED file"

python $read_distribution -i $file -r $RefBED 1>>readDistribution.output.txt 2>/dev/null
```

## Output of read_distribution.py

```{r Output read distribution, echo=FALSE}
cat(readLines("C:/Users/lingzili/Documents/seq_tutorials/data/SM4998.readDist.txt"), sep = "
")
```

## Gene body coverage

Calculate the RNA-seq reads coverage over gene body. 

```{r example geneBodyCoverage, echo=FALSE, eval=TRUE, out.width = "600px"}
knitr::include_graphics("C:/Users/lingzili/Documents/seq_tutorials/plot/Example_geneBodyCoverage.png")
```

## geneBody_coverage.py

```{bash Genebody coverage, eval=FALSE}
locate geneBody_coverage
geneBody_coverage="your file path to geneBody_coverage.py"

housekeepBED="your file path to housekeeping gene BED file"

# Only input sorted and indexed BAM files
file1="your BAM file path"
file2="your BAM file path"
file3="etc. etc."

prefix_output="prefix of your output"

python $geneBody_coverage -r $housekeepBED -i $file1,$file2 -o $prefix_output
```

## Output of geneBody_coverage.py

- The line plots should look like this: "∩". 
- Using oligo-dT beads for mRNA-seq prep can be prone to 3' bias in genebody coverage if degraded RNA (RIN < 7) is used as input, which may introduce quantification bias.
```{r plot geneBodyCoverage, echo=FALSE, eval=TRUE, out.width = "400px"}
knitr::include_graphics("C:/Users/lingzili/Documents/seq_tutorials/plot/geneBodyCoverage.png")
```

## Supplementary information

```{bash Ref gene download, eval=FALSE}
# Download mm10 rRNA BED file
wget https://iweb.dl.sourceforge.net/project/rseqc/BED/Mouse_Mus_musculus/mm10_rRNA.bed.gz
gunzip mm10_rRNA.bed.gz
# Download mm10 Ref genome BED file
wget https://iweb.dl.sourceforge.net/project/rseqc/BED/Mouse_Mus_musculus/mm10_RefSeq.bed.gz
gunzip mm10_RefSeq.bed.gz
# Donwload mm10 housekeeping BED file
wget https://iweb.dl.sourceforge.net/project/rseqc/BED/Mouse_Mus_musculus/mm10.HouseKeepingGenes.bed.gz
gunzip mm10.HouseKeepingGenes.bed.gz

# Download human geneome BED files, see: https://sourceforge.net/projects/rseqc/files/BED/
```

To get the most recent annotation, use UCSC's Table Browser <https://genome.ucsc.edu/cgi-bin/hgTables?command=start>

## Bibliography

- Liguo Wang, Shengqin Wang, Wei Li (2012). RSeQC: quality control of RNA-seq experiments. Bioinformatics, 28(16), 2184–2185. http://doi.org/10.1093/bioinformatics/bts356

- “Picard Toolkit.” 2019. Broad Institute, GitHub Repository. http://broadinstitute.github.io/picard/

- Philip Ewels, Måns Magnusson, Sverker Lundin and Max Käller (2016). MultiQC: Summarize analysis results for multiple tools and samples in a single report. Bioinformatics, 32(19), 3047–3048. https://doi.org/10.1093/bioinformatics/btw354 

- Simon Andrews (2010). FastQC: a quality control tool for high throughput sequence data. Available online at: http://www.bioinformatics.babraham.ac.uk/projects/fastqc
