---
title: "RSeQC demo"
author: "JWK & Islet group coding club meet-up"
date: "July 17, 2019"
output:
  ioslides_presentation: default
  powerpoint_presentation: default
---
<style type="text/css">

body{ /* Normal  */
      font-size: 16px;
  }
td {  /* Table  */
  font-size: 12px;
}
code.r{ /* Code block */
    font-size: 14px;
}
pre { /* Code block - determines code spacing between lines */
    font-size: 14px;
}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Quality control checkpoints

QC should be applied to every step of RNA-seq: raw sequencing reads, alignment and quantification. 
The tools involved are:

- **FastQC** <https://www.bioinformatics.babraham.ac.uk/projects/fastqc/>
- **MultiQC** <https://multiqc.info/>
- **RSeQC** <http://rseqc.sourceforge.net/>
- **Picard**

## Output of aligner: SAM/BAM
The SAM (Sequence Alignment Map) file contains information for each individual read and its alignment to the genome. BAM is a binary, compressed version of the SAM file.

|No.	|Name	|Description|
|----:|:----|-----------|
|1	|QNAME|	Query NAME of the read or the read pair|
|2	|FLAG|	Bitwise FLAG (pairing, strand, mate strand, etc.)|
|3	|RNAME|	Reference sequence NAME|
|4	|POS|	1-Based leftmost POSition of clipped alignment|
|5	|MAPQ|	MAPping Quality (Phred-scaled)|

## Output of aligner: SAM/BAM

|No.	|Name	|Description|
|----:|:----|-----------|
|6	|CIGAR|	Extended CIGAR string (operations: MIDNSHP)|
|7	|MRNM|	Mate Reference NaMe (‘=’ if same as RNAME)|
|8	|MPOS|	1-Based leftmost Mate POSition|
|9	|ISIZE|	Inferred Insert SIZE|
|10	|SEQ|	Query SEQuence on the same strand as the reference|
|11	|QUAL|Query QUALity (ASCII-33=Phred base quality)|

## Evaluation of RNA integrity at transcript level

```{bash echo = TRUE, eval = FALSE}
tin.py -i *.bam -r ref.bed -o OUTPUT_PREFIX
```
