---
title: "Transcript Quantification"
author: "JWK & Islet group coding club meet-up"
date: "7/11/2019"
output: slidy_presentation
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

## Quantification metrics

- Normalization for comparison within a **single** sample
- Normalization for comparison among **multiple** samples/conditions

You should **never** use RPKM, FPKM, or TPM to compare expression levels across samples, because they are **relative** measures. 

## GTF (General Feature Format)

Contains 9 columns of gene/transcript related features: **seqname, source, feature (e.g. Gene, Exon), start, end, score, strand, frame, attribute (e.g. gene_id, gene_name)**

Example: 

1       ensembl_havana  gene    5588466 5606131 .       +       .       gene_id "ENSMUSG00000025905"; gene_version "9"; gene_name "Oprk1"; gene_source "ensembl_havana"; gene_biotype "protein_coding";
