---
title: "kallisto demo"
author: "JWK & Islet group coding club meet-up"
date: "July 30, 2019"
output:
  ioslides_presentation: default
  powerpoint_presentation: default
---
<style type="text/css">
code.r{
  font-size: 14px;
}
pre {
  font-size: 14px
}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
library(knitr)
library(scales)
```

## Basic concept: k-mer

k-mer typically refers to all the possible substrings of length k that are contained in a string.

```{r k-mer, echo=FALSE}
# Define a function to generate k-mers
k_mers <- function(sequence, k) {
  substring(sequence, 1:(nchar(sequence) - k + 1), k:nchar(sequence))
}

# Define input read
input_read <- "TACCGA"
cat(sprintf("input read = %s\n", input_read))

# Print all possible k-mers
for (k in 1:nchar(input_read)) {
  cat(sprintf("k-mer = %s\n", k))
  print(k_mers(sequence = input_read, k))
}
```

## Download mouse transcriptome reference

Download FASTA file for cDNAs of mouse (GRCm38.p6) from Ensembl
<http://www.ensembl.org/Mus_musculus/Info/Index>

```{bash echo = TRUE, eval = FALSE}
mkdir kallisto_demo
cd kallisto_demo

# Download mouse cDNA FASTA file
wget ftp://ftp.ensembl.org/pub/release-97/fasta/mus_musculus/cdna/Mus_musculus.GRCm38.cdna.all.fa.gz
```

## FASTA

Starts with ‘>’ and gives basic information about sequence

Example:

\>4 ENSMUST00000070533 chr1- 3214482-3216968,3421702-3421901,3670552-3671498
AAAGTTTGACCGGCACTTTTAATGCAACAAAAAGCACCATGACCAAGGACACCGAAAACA
GAAAGCGGGAATGTTTACTTGCTGTGTGGAGGTATTTAAAACTCAGAAGCTTGTAAGAGC

## Start with kallisto

To check the version of kallisto, type in command line:
```{bash echo = TRUE, eval = FALSE}
kallisto version
```

To look for general help, type:
```{bash echo = TRUE, eval = FALSE}
kallisto
```

To look for specific help, for example on building an index, type:
```{bash echo = TRUE, eval = FALSE}
kallisto index
```

## Build “transcriptome index”

Input: Reference transciptome; output: index file

```{bash echo = TRUE, eval = FALSE}
kallisto index -i GRCm38.cdna.all.idx Mus_musculus.GRCm38.cdna.all.fa.gz
```

## Quantification of transcript abundances

In the "quant" function we specify:

- number of threads to use (parameter "-t");
- location of the trasncriptome index (parameter "-i");
- kallisto output folder for each dataset (parameter "-o");
- number of bootstrap samples to estimate the technical variance (parameter "-b");

and finally, the location of the paired-end fastq, compressed or uncompressed.

Note: For **single-end reads**, the parameters are different and please check the kallisto manual!

## Quantification of transcript abundances

Pair-end reads:
```{bash echo = TRUE, eval = FALSE}
kallisto quant -t 4 \
-i GRCm38.cdna.all.idx \
-o kallisto_<read_id> \
-b 100 \
reads_1.fastq.gz reads_2.fastq.gz
```

Note: **Bootstrapping** is the process of repeating the quantification analysis after resampling with replacement from the original data, in order to simulate random sampling associated with sequencing. In general, this number should be **at least 30**.

## What's next?

**Sleuth**: A differential expression tool that accepts directly the outputs of kallisto (see: <https://pachterlab.github.io/sleuth/>)

## Caution: When to use kallisto

Douglas C. Wu, Jun Yao, Kevin S. Ho, Alan M. Lambowitz and Claus O. Wilke, [Limitations of alignment-free tools in total RNA-seq quantification](https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-018-4869-5#Sec4), BMC Genomics 19, Article number: 510 (2018)

*"Accurate quantification of lowly-expressed or small RNA is difficult to achieve with alignment-free pipelines"*

## Bibliography
kallisto, Pachter Lab <https://pachterlab.github.io/kallisto/>

Nicolas L Bray, Harold Pimentel, Páll Melsted and Lior Pachter, [Near-optimal probabilistic RNA-seq quantification](https://www.nature.com/articles/nbt.3519), Nature Biotechnology 34, 525–527 (2016)
